libminifi/src/c2/C2Agent.cpp

```diff
@@ -90,22 +87,16 @@ C2Agent::C2Agent(const std::shared_ptr<core::controller::ControllerServiceProvid
         request_mutex.unlock();
       }
 
-      if ( time_since > heart_beat_period_ ) {
-        last_run_ = now;
-        try {
-          performHeartBeat();
-        }
-        catch(const std::exception &e) {
-          logger_->log_error("Exception occurred while performing heartbeat. error: %s", e.what());
-        }
-        catch(...) {
-          logger_->log_error("Unknonwn exception occurred while performing heartbeat.");
-        }
+      try {
+        performHeartBeat();
+      } catch(const std::exception &e) {
+        logger_->log_error("Exception occurred while performing heartbeat. error: %s", e.what());
+      } catch(...) {
+        logger_->log_error("Unknonwn exception occurred while performing heartbeat.");
       }
 
       checkTriggers();
 
-      std::this_thread::sleep_for(std::chrono::milliseconds(heart_beat_period_ > 500 ? 500 : heart_beat_period_));
```

Comment:
Do we honor the heartbeat period here? I mean doesn't the threadpool schedule this task immediately or at a different delay interval that the heartbeat period.
Also, I believe this PR will conflict with MINFICPP-1169. In that PR we are using the new theadpool changes.

Reply:
Yes we do.

It's done in this call:
```
registerUpdateListener(agent, agent->getHeartBeatDelay());
```

Which creates the afterexecute based on the heartbeat delay:

```
std::unique_ptr<utils::AfterExecute<Update>> after_execute = std::unique_ptr<utils::AfterExecute<Update>>(new UpdateRunner(isStateMonitorRunning(), delay));
```

So this works fine without any additional hack inside the callback.
I also based it on master. 

If you want to fix it any other way in 1169, I'm fine  with that, just make sure heartbeat timing is the responsibility of threadpool scheduling and not some blackmagic done in the callbacks.

Reply:
This is how it is done in 1169 if that is what you mean by letting threadpool to take the scheduling.
     try {
        performHeartBeat();
      }
      catch(const std::exception &e) {
        logger_->log_error("Exception occurred while performing heartbeat. error: %s", e.what());
      }
      catch(...) {
        logger_->log_error("Unknonwn exception occurred while performing heartbeat.");
      }

      checkTriggers();
      return utils::TaskRescheduleInfo::RetryIn(std::chrono::milliseconds(heart_beat_period_));